File: src/utils/chartData.ts
import { getCampaigns, computeCampaignROI, getCampaignMetrics } from '../data/dataStore';
import type { Campaign } from '../types';

export interface ChartPoint {
  label: string;
  value: number;
}

export interface ChartSegment {
  label: string;
  value: number;
  color: string;
}

export const CHART_COLORS = ['#2b8fff', '#38c8ff', '#1470eb', '#7ddcff', '#0cb0f5', '#1048a3', '#0d58c7', '#0090d4'];

export function getCampaignSpendBars(campaigns: Campaign[]): { label: string; value: number; max: number }[] {
  return campaigns.map((c) => ({
    label: c.name.length > 14 ? `${c.name.slice(0, 14)}…` : c.name,
    value: c.spent,
    max: c.budget || c.spent || 1,
  }));
}

export function getChannelSegments(campaigns: Campaign[]): ChartSegment[] {
  const map: Record<string, number> = {};
  campaigns.forEach((c) => {
    c.channels.forEach((ch) => { map[ch] = (map[ch] ?? 0) + 1; });
  });
  return Object.entries(map).map(([label, value], i) => ({
    label,
    value,
    color: CHART_COLORS[i % CHART_COLORS.length],
  }));
}

export function getStatusSegments(campaigns: Campaign[]): ChartSegment[] {
  const map: Record<string, number> = {};
  campaigns.forEach((c) => { map[c.status] = (map[c.status] ?? 0) + 1; });
  return Object.entries(map).map(([label, value], i) => ({
    label,
    value,
    color: CHART_COLORS[i % CHART_COLORS.length],
  }));
}

export function getRoiBars(campaigns: Campaign[]): ChartPoint[] {
  return campaigns.map((c) => ({
    label: c.name.length > 10 ? `${c.name.slice(0, 10)}…` : c.name,
    value: computeCampaignROI(c).roi,
  }));
}

export function getSpendTrend(campaigns: Campaign[]): ChartPoint[] {
  const map: Record<string, number> = {};
  campaigns.forEach((c) => {
    const month = new Date(c.startDate).toLocaleString('en-PH', { month: 'short' });
    map[month] = (map[month] ?? 0) + c.spent;
  });
  const entries = Object.entries(map);
  if (entries.length === 0) {
    return ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun'].map((label, i) => ({
      label,
      value: Math.round(campaigns.reduce((s, c) => s + c.spent, 0) * (0.08 + i * 0.04)),
    }));
  }
  return entries.map(([label, value]) => ({ label, value }));
}

export function getReachBars(campaigns: Campaign[]): ChartPoint[] {
  return campaigns.map((c) => {
    const m = getCampaignMetrics(c.id);
    return {
      label: c.name.length > 10 ? `${c.name.slice(0, 10)}…` : c.name,
      value: m?.reach ?? 0,
    };
  });
}

export function getBudgetUtilization(campaigns: Campaign[]): number {
  const budget = campaigns.reduce((s, c) => s + c.budget, 0);
  const spent = campaigns.reduce((s, c) => s + c.spent, 0);
  return budget > 0 ? Math.round((spent / budget) * 100) : 0;
}

export function getAvgRoi(campaigns: Campaign[]): number {
  if (campaigns.length === 0) return 0;
  return Math.round(campaigns.reduce((s, c) => s + computeCampaignROI(c).roi, 0) / campaigns.length);
}

export function getAllCampaigns() {
  return getCampaigns();
}

File: mobiledevelopment/lib/src/utils/chartData.dart
import 'package:flutter/material.dart';
import '../data/dataStore.dart';
import '../types/index.dart';

class ChartPoint {
  final String label;
  final double value;
  const ChartPoint({required this.label, required this.value});
}

class ChartSegment {
  final String label;
  final double value;
  final String color;
  const ChartSegment({required this.label, required this.value, required this.color});
}

class BarChartItem {
  final String label;
  final double value;
  final double max;
  const BarChartItem({required this.label, required this.value, required this.max});
}

const chartColors = ['#2b8fff', '#38c8ff', '#1470eb', '#7ddcff', '#0cb0f5', '#1048a3', '#0d58c7', '#0090d4'];

Color hexToColor(String hex) {
  final h = hex.replaceFirst('#', '');
  return Color(int.parse('FF$h', radix: 16));
}

List<BarChartItem> getCampaignSpendBars(List<Campaign> campaigns) {
  return campaigns.map((c) => BarChartItem(
        label: c.name.length > 14 ? '${c.name.substring(0, 14)}…' : c.name,
        value: c.spent,
        max: c.budget > 0 ? c.budget : (c.spent > 0 ? c.spent : 1),
      )).toList();
}

List<ChartSegment> getChannelSegments(List<Campaign> campaigns) {
  final map = <String, int>{};
  for (final c in campaigns) {
    for (final ch in c.channels) {
      map[ch] = (map[ch] ?? 0) + 1;
    }
  }
  return map.entries.toList().asMap().entries.map((e) => ChartSegment(
        label: e.value.key,
        value: e.value.value.toDouble(),
        color: chartColors[e.key % chartColors.length],
      )).toList();
}

List<ChartSegment> getStatusSegments(List<Campaign> campaigns) {
  final map = <String, int>{};
  for (final c in campaigns) {
    map[c.status] = (map[c.status] ?? 0) + 1;
  }
  return map.entries.toList().asMap().entries.map((e) => ChartSegment(
        label: e.value.key,
        value: e.value.value.toDouble(),
        color: chartColors[e.key % chartColors.length],
      )).toList();
}

List<ChartPoint> getRoiBars(List<Campaign> campaigns) {
  return campaigns.map((c) {
    final roi = dataStore.computeCampaignROI(c).roi;
    return ChartPoint(
      label: c.name.length > 10 ? '${c.name.substring(0, 10)}…' : c.name,
      value: roi.toDouble(),
    );
  }).toList();
}

List<ChartPoint> getSpendTrend(List<Campaign> campaigns) {
  final map = <String, double>{};
  for (final c in campaigns) {
    final month = _monthShort(c.startDate);
    map[month] = (map[month] ?? 0) + c.spent;
  }
  if (map.isEmpty) {
    final total = campaigns.fold<double>(0, (s, c) => s + c.spent);
    return ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun'].asMap().entries.map((e) => ChartPoint(label: e.value, value: (total * (0.08 + e.key * 0.04)).roundToDouble())).toList();
  }
  return map.entries.map((e) => ChartPoint(label: e.key, value: e.value)).toList();
}

List<ChartPoint> getReachBars(List<Campaign> campaigns) {
  return campaigns.map((c) {
    final m = dataStore.getCampaignMetrics(c.id);
    return ChartPoint(
      label: c.name.length > 10 ? '${c.name.substring(0, 10)}…' : c.name,
      value: (m?['reach'] as num?)?.toDouble() ?? 0,
    );
  }).toList();
}

int getBudgetUtilization(List<Campaign> campaigns) {
  final budget = campaigns.fold<double>(0, (s, c) => s + c.budget);
  final spent = campaigns.fold<double>(0, (s, c) => s + c.spent);
  return budget > 0 ? ((spent / budget) * 100).round() : 0;
}

int getAvgRoi(List<Campaign> campaigns) {
  if (campaigns.isEmpty) return 0;
  final total = campaigns.fold<int>(0, (s, c) => s + dataStore.computeCampaignROI(c).roi);
  return (total / campaigns.length).round();
}

String _monthShort(String date) {
  try {
    final d = DateTime.parse(date);
    const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
    return months[d.month - 1];
  } catch (_) {
    return 'N/A';
  }
}
